### PR TITLE
Allow participants to opt-out and opt-in for data collection during a testday

### PR DESCRIPTION
**Pull request by galgeek**
Hi @whimboo!

This PR renames and removes old stats variables and enables opt out from nick-identified data collection. Opt out totals are added to stats. This PR addresses issues #44 and #23 (privacy issues).

It also includes a couple of fixes for rebase errors.

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -34,7 +34,8 @@ var ircServer = config.server,
              ":sumo" : "Learn about Support at Mozilla",
              ":etherpad" : "View the Test Day etherpad",
              ":helpers" : "View Test Day helpers, and request help with :helpers request",
-             ":schedule" : "View the Test Day schedule"
+             ":schedule" : "View the Test Day schedule",
+             ":optout"   : "Opt out from Test Day data collection for your nick"
```
Comment (whimboo) [inline bot.js]: People may reconsider their decision. So we should have a way for opt-in again.

File: bot.js [hunk 2]
```diff
@@ -168,17 +157,36 @@ client.addListener('message', function(from, to, message) {
     client.say(to, intro + scheduleTimes);
   }
 
+  if (message.search('[!:]optout') === 0) {
+    if (!(from in optOut)) {
+      optOut[from] = null;
```
Comment (whimboo) [inline bot.js]: Do we really need a hash? Otherwise it can be a plain array.

File: bot.js [hunk 3]
```diff
@@ -168,17 +157,36 @@ client.addListener('message', function(from, to, message) {
...
+      metrics.optOutTotal += 1;
+      if (from in metrics.activeUsers) {
+        delete metrics.activeUsers[from];
+      }
+    }
+    client.say(from, "You’ve opted out of Test Day data collection " +
+    "for your nick " + from + ".");
```
Comment (whimboo) [inline bot.js]: nit: the second line should be aligned with the first parameter of the method.

File: bot.js [hunk 4]
```diff
@@ -168,17 +157,36 @@ client.addListener('message', function(from, to, message) {
...
-      metrics.usersTalked[from] = 1;
+
+    // if from is not on the opt out list
+    if (!(from in optOut)) {
+      if (from in metrics.activeUsers) {
+        metrics.activeUsers[from] += 1;
+      } else {
+        client.say(from, "Welcome to today’s Test Day, " + from + "!");
+          client.say(from, "To opt out of data collection, use the command :optout.");
+          metrics.activeUsers[from] = 1;
```
Comment (whimboo) [inline bot.js]: nit: please fix the indentation.

File: bot.js [hunk 5]
```diff
@@ -258,7 +265,7 @@ client.addListener('pm', function(from, message) { // private messages to bot
               if (timerID !== 0) {
                 clearTimeout(timerID);
               }
-              timerID = setTimeout(checkTestDay, startTime - Date.now());
+              timerID = setTimeout(updateTestDayData, startTime - Date.now());
```
Comment (whimboo) [inline bot.js]: Those two changes in `pm` where already part of another PR, which has been merged. Please double check in the future that you do not duplicate changes between PRs.

File: bot.js (hunk 2 above)
Comment (galgeek) [inline bot.js]: Finding in a hash is supposed to be faster than finding in a plain array.
I have no idea how big the optout list may become, and it's internal, so it doesn't need to be easy to edit in a config file, for instance.

**Comment by galgeek** [comment]
Hi @whimboo!

I've updated per your comments. On to saving some data to disk...

Thank you!

File: bot.js (hunk 2 above)
Comment (whimboo) [inline bot.js]: Well, the question here is if we want to store more information on such a user listed in this array. If we don't do that, there is no reason to reserve extra memory for a null value.

Comment (galgeek) [inline bot.js]: I'd say there's nothing more to store for opt-outs, so if we're concerned about memory, it'd be good to rewrite as array.

**Comment by galgeek** [comment]
Hi @whimboo!

I've updated this PR:

* I've rewritten optOut as an array. 
* In a separate commit, I've removed lastQuit{} and the quit and part listeners handling it, since we're no longer greeting based on that. This will also save some memory.

**Comment by galgeek** [comment]
Hello @whimboo!

I've updated again... the bot can store the optout list across restarts.

Perhaps you'd prefer to include this in a separate PR?

**Comment by whimboo** [comment]
> I've updated again... the bot can store the optout list across restarts.
> Perhaps you'd prefer to include this in a separate PR?

This is great but as I have told yesterday already, it has to be a separate issue and PR. So please move this out. I will wait with the review.

**Comment by whimboo** [comment]
Also keep in mind that we have to store most of the data in a file so it survives a restart. It will be best to combine that all in the necessary issue.

**Comment by galgeek** [comment]
Hi, @whimboo!

I've updated, backing out all the files code for a separate PR.

File: bot.js [hunk 6]
```diff
@@ -185,17 +173,48 @@ client.addListener('message', function(from, to, message) {
...
+      metrics.optOutTotal += 1;
+      if (from in metrics.activeUsers) {
+        delete metrics.activeUsers[from];
+      }
+    }
+    client.say(from, "You’ve opted out of Test Day data collection " +
+               "for your nick " + from + ".");
+  }
+
+  if (message.search('[!:]optin') === 0) {
```
Comment (whimboo) [inline bot.js]: Please file a follow-up issue so we make use of `elif` or even better a switch construct for the command. At the moment we are doing all the if checks even the first if block has the command.

File: bot.js [hunk 7]
```diff
@@ -185,17 +173,48 @@ client.addListener('message', function(from, to, message) {
...
       if (message.search(/https:\/\/bugzilla.mozilla.org\/show_bug.cgi\?id=(\d+)/i) >= 0) {
         metrics.firebotBugs.push(/https:\/\/bugzilla.mozilla.org\/show_bug.cgi\?id=(\d+)/i.exec(message)[1]);
       }
     }
-    if (from in metrics.usersTalked) {
-      metrics.usersTalked[from] += 1;
-    } else {
-      metrics.usersTalked[from] = 1;
+
+    // if from is not on the opt out list
```
Comment (whimboo) [inline bot.js]: I think the comment should say that this part is about offering the opt-out option.

Comment (whimboo) [inline bot.js]: Wait there is even more in this block. So just update the comment to make it clear what this block is about.

Comment (whimboo) [inline bot.js]: Well, lets do it later given that everything else looks fine and I can get this merged.

**Comment by whimboo** [comment]
I squashed the commits, updated the commit message, and merged the PR as https://github.com/mozilla/testdaybot/commit/c94db45aaedf9d71af8d7720b014892bef395cd3

**Pull request closed without merging**